File: python/time_needed_to_inform_all_employees.py
from collections import deque, defaultdict


class Solution:
    def numOfMinutes(
        self, n: int, headID: int, manager: list[int], informTime: list[int]
    ) -> int:
        """time complexity: O(n) by using bfs"""
        adj: dict[int, list[int]] = {}
        for i in range(n):
            if manager[i] not in adj:
                adj[manager[i]] = [i]
            else:
                adj[manager[i]].append(i)
        que: deque[tuple[int, int]] = deque([(headID, informTime[headID])])
        result: int = 0
        while len(que) > 0:
            i, time = que.popleft()
            result = max(result, time)
            if i not in adj:
                continue
            for employ in adj[i]:
                que.append((employ, time + informTime[employ]))
        return result

    def numOfMinutes_dfs(
        self, n: int, headID: int, manager: list[int], informTime: list[int]
    ) -> int:
        """time complexity: O(n) by using dfs"""
        adj: dict[int, list[int]] = defaultdict(list)
        for i in range(n):
            adj[manager[i]].append(i)
        result: list[int] = [0]
        self.dfs(headID, 0, result, adj, manager, informTime)
        return result[0]

    def dfs(
        self,
        i: int,
        accu_time: int,
        result: list[int],
        adj: dict[int, list[int]],
        manager: list[int],
        informTime: list[int],
    ) -> None:
        if informTime[i] == 0:
            return
        accu_time += informTime[i]
        result[0] = max(result[0], accu_time)
        for employee in adj[i]:
            self.dfs(employee, accu_time, result, adj, manager, informTime)


def test_numOfMinutes_case_1():
    """This is a test case"""
    # arrange
    n: int = 1
    headID: int = 0
    manager: list[int] = [-1]
    informTime: list[int] = [0]
    expected: int = 0

    # act
    solution = Solution()
    actual = solution.numOfMinutes(n, headID, manager, informTime)

    # assert
    assert expected == actual


def test_numOfMinutes_case_2():
    """This is a test case"""
    # arrange
    n: int = 6
    headID: int = 2
    manager: list[int] = [2, 2, -1, 2, 2, 2]
    informTime: list[int] = [0, 0, 1, 0, 0, 0]
    expected: int = 1

    # act
    solution = Solution()
    actual = solution.numOfMinutes(n, headID, manager, informTime)

    # assert
    assert expected == actual


def test_numOfMinutes_case_3():
    """This is a test case"""
    # arrange
    n: int = 6
    headID: int = 2
    manager: list[int] = [2, 2, -1, 2, 2, 2]
    informTime: list[int] = [0, 0, 1, 0, 0, 0]
    expected: int = 1

    # act
    actual = Solution().numOfMinutes_dfs(n, headID, manager, informTime)

    # assert
    assert expected == actual


def test_numOfMinutes_case_4():
    """This is a test case"""
    # arrange
    n: int = 1
    headID: int = 0
    manager: list[int] = [-1]
    informTime: list[int] = [0]
    expected: int = 0

    # act
    actual = Solution().numOfMinutes_dfs(n, headID, manager, informTime)

    # assert
    assert expected == actual

File: rust/src/code/time_needed_to_inform_all_employees.rs
struct Solution;

use std::collections::{HashMap, VecDeque};

impl Solution {
    pub fn num_of_minutes_bfs(
        n: i32,
        head_id: i32,
        manager: Vec<i32>,
        inform_time: Vec<i32>,
    ) -> i32 {
        let mut result: i32 = 0;
        let mut adj: HashMap<usize, Vec<usize>> = HashMap::new();
        for i in 0..n {
            adj.entry(manager[i as usize] as usize)
                .or_insert(vec![])
                .push(i as usize);
        }
        let mut queue: VecDeque<(usize, i32)> =
            VecDeque::from(vec![(head_id as usize, inform_time[head_id as usize])]);
        while let Some((node, time)) = queue.pop_front() {
            result = std::cmp::max(result, time);
            if !adj.contains_key(&node) {
                continue;
            }
            for employee in adj[&node].iter() {
                queue.push_back((*employee, time + inform_time[*employee]));
            }
        }
        result
    }

    // Time complexity: O(n)
    pub fn num_of_minutes(n: i32, head_id: i32, manager: Vec<i32>, inform_time: Vec<i32>) -> i32 {
        let mut result: Vec<i32> = vec![0];
        let mut adj: HashMap<usize, Vec<usize>> = HashMap::new();
        for i in 0..n {
            adj.entry(manager[i as usize] as usize)
                .or_insert(vec![])
                .push(i as usize);
        }
        Self::dfs(head_id as usize, 0, &mut result, &adj, &inform_time);
        result[0]
    }

    fn dfs(
        i: usize,
        accu_time: i32,
        result: &mut Vec<i32>,
        adj: &HashMap<usize, Vec<usize>>,
        inform_time: &Vec<i32>,
    ) {
        if inform_time[i] == 0 {
            return;
        }
        result[0] = std::cmp::max(result[0], accu_time + inform_time[i]);
        for employee in adj[&i].iter() {
            Self::dfs(
                *employee,
                accu_time + inform_time[i],
                result,
                adj,
                inform_time,
            );
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_num_of_minutes_case_1() {
        // arrange
        let n: i32 = 1;
        let head_id: i32 = 0;
        let manager: Vec<i32> = vec![-1];
        let inform_time: Vec<i32> = vec![0];
        let expected: i32 = 0;

        // act
        let mut actual = Solution::num_of_minutes(n, head_id, manager, inform_time);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_num_of_minutes_case_2() {
        // arrange
        let n: i32 = 6;
        let head_id: i32 = 2;
        let manager: Vec<i32> = vec![2, 2, -1, 2, 2, 2];
        let inform_time: Vec<i32> = vec![0, 0, 1, 0, 0, 0];
        let expected: i32 = 1;

        // act
        let mut actual = Solution::num_of_minutes(n, head_id, manager, inform_time);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_num_of_minutes_case_3() {
        // arrange
        let n: i32 = 6;
        let head_id: i32 = 2;
        let manager: Vec<i32> = vec![2, 2, -1, 2, 2, 2];
        let inform_time: Vec<i32> = vec![0, 0, 1, 0, 0, 0];
        let expected: i32 = 1;

        // act
        let mut actual = Solution::num_of_minutes_bfs(n, head_id, manager, inform_time);

        // assert
        assert_eq!(expected, actual);
    }
}
